### PR TITLE
fix(train): report unavailable loss for MLX and unlearn

### LINKED ISSUE #960: mlx_sft and unlearn still report Loss: 0.0000 where there was no measurement
#908 replaced the fake loss delta with a `loss_summary` helper and migrated the trainers that use HF `log_history`. Two trainers were deliberately left outside it, and they are outside its regression guard too.

`trainer/mlx_sft.py:770-773` and `trainer/unlearn.py:320-321` still build `{"initial_loss", "final_loss"}` by hand with a `0.0` fallback:

```python
losses = captured.get("losses") or [0.0]     # mlx_sft.py:770
```

Neither is in `_MIGRATED_TRAINERS` (`tests/test_issue899_loss_summary.py:8-27`).

### What is and is not already fixed on those backends

The `summary_kind is None and initial == final` fallback at `commands/train.py:75-77` **does** remove the fake arrow, so #899's headline symptom — `Loss: 2.3830 -> 2.3830` presented as a delta — is gone on MLX and unlearn too.

What remains: a run where no loss was captured prints

```
Loss: 0.0000
```

rather than `unavailable`. A plausible number where there is no measurement is the failure mode this project keeps getting bitten by, and it is worse than an error because nobody investigates it. And nothing guards a regression there — mutating either fallback changes no test.

### Why they were left out, and why that was reasonable

Neither backend uses HF `log_history`, so they cannot share the migrated code path as-is. Leaving them out kept #908 reviewable, which was the right call for that PR. This issue is the other half, not a criticism of it.

### Acceptance

- [ ] A run on either backend with no captured losses reports `unavailable`, not `0.0000`
- [ ] Both trainers appear in `_MIGRATED_TRAINERS`, or the table grows a second list for backends that report a summary without `log_history`
- [ ] A mutation of each fallback fails a named test — the anchor counted exactly once, the module still importing, and the failure a behavioural assert rather than a collection error
- [ ] `initial_loss` / `final_loss` keep their numeric meaning for the tracker and for `soup sweep` ranking, as documented at `loss_summary.py:29-36`

Follow-up to #899, closed by #908 by @taljeon.

### PR DESCRIPTION
**Pull request by Rayan-and-beyond**
## Summary

- route MLX SFT and unlearn completion losses through the shared loss-summary helper
- report `unavailable` instead of inventing `0.0000` when neither backend captured a loss
- report one-step unlearn runs as `single` rather than a fake delta
- add behavioral regressions through the real Unlearn and fake-MLX-backed training paths

Closes #960

## Validation

- `tests/test_issue899_loss_summary.py` — 18 passed on Python 3.11
- loss-summary + MLX backend-support focused suites — 68 passed
- reintroduced the old `or [0.0]` fallback separately in Unlearn and MLX — each new no-loss behavioral test failed
- `ruff check` on the touched Python files — passed
- `git diff --check` — passed

### CODE REVIEW

**Comment by MakazhanAlpamys** [comment]
**Your CI went red about an hour ago and it is almost certainly not your diff.**

`peft 0.21.0` was published to PyPI at **13:33:59Z today**. `pyproject.toml` declares `peft>=0.20.0,<1.0.0`, so CI took it on the next run, and **nine layer-streaming adapter tests now fail on every PR that has run since** — `tests/test_v07201.py`, `tests/test_v07202.py`, `tests/test_v07203.py`.

Tracked as **#1005**, where the root cause is now established and measured: it is a **real regression**, not a flaky test. peft 0.21 changed adapter-entry selection from string-matching `state_dict` keys to a structural walk of `named_modules()`, and Soup's streaming wrapper has a canonical `state_dict()` over a `.inner.` module tree — so the intersection is empty and a streamed run writes a **0-tensor adapter with no exception raised**.

**What this means for you:** if your failures are in those three files and your diff touches none of `trainer/`, `utils/layer_stream*` or PEFT, there is nothing to do. **Do not rebase or force-push chasing it** — a new run will fail the same way until the fix lands.

If you see a failure *outside* those three files, that one is worth a look, and please say so here rather than assuming it is this.

A fix is in progress on my side (not a version pin — the pin would hide the defect from anyone who installs a newer peft themselves). I will ping this thread when `main` is green again.

Sorry for the noise.

**Comment by MakazhanAlpamys** [comment]
**`main` is green again** — #1010 merged as `f8226214`, and I have requested a CI re-run on this PR.

As promised when I told you the red was not yours: it was `peft 0.21.0` changing how it selects adapter tensors, which made every streamed LoRA run save a 0-tensor adapter. Details and measurements are on #1005.

**Nothing for you to do.** `pull_request` workflows build against the merge ref, so the re-run picks up the new `main` without moving your head SHA — no rebase, no force-push, and any review verdict recorded against your head still stands.

If it comes back red and the failures are **not** in `tests/test_v07201.py` / `test_v07202.py` / `test_v07203.py`, that one is worth a look and is not this. Say so here and I will classify it rather than guess.

**Comment by MakazhanAlpamys** [comment]
**Correcting my comment above. I told you "nothing for you to do" and that was wrong — there is one thing, and without it this PR stays red indefinitely.**

I wrote:

> `pull_request` workflows build against the merge ref, so the re-run picks up the new `main` without moving your head SHA — no rebase, no force-push

The first half is right and the conclusion does not follow. **The merge ref is pinned when the run is created, not resolved when the job starts.** From an actual job log on another PR:

```
git fetch --depth=1 origin +8378a316...:refs/remotes/pull/763/merge
git checkout --force refs/remotes/pull/763/merge
HEAD is now at 8378a31 Merge <head> into 48ea2147ffad9df21a6d8c1d13a77bef9cd9754d
```

The fetch names an **explicit SHA**, so the ref is not recomputed, and the base it merged into is `48ea2147` — 15 commits behind today's `main`, and `git merge-base --is-ancestor f8226214 8378a316` confirms it does **not** contain #1010.

That job acquired its runner *after* #1010 merged and still checked out the pre-fix tree. So when a job runs is irrelevant; only when its run was **created** matters. A re-run re-runs the job against the same frozen merge and cannot produce a different answer.

## What actually clears it

A **new event** on the PR, which is the only thing that recomputes the merge commit:

1. **You push** — a rebase onto current `main`, or any commit; or
2. **I press "Update branch"** (`gh pr update-branch`), merging current `main` into your branch. Your commits and authorship are untouched; it adds one merge commit. `--rebase` is available if you prefer that shape.

**Tell me which and I will do it.** I will not write to your branch without you saying so. If you would rather push yourself, do it whenever suits — nothing is lost by waiting, and any review verdict recorded against your current head still stands.

## What has not changed

The diagnosis was right and none of this is charged to you. The nine `test_v07201.py` / `test_v07202.py` / `test_v07203.py` failures are #1005 — `peft 0.21.0` changing how it selects adapter tensors, so every streamed LoRA run saved a 0-tensor adapter — fixed by #1010 at 2026-09-16T05:37:30Z. Your branch is not involved.

And the rest of that comment stands: if it comes back red with failures **outside** those three files, that one is worth a look and is not this. Say so here and I will classify it rather than guess.

## Why I am posting this rather than quietly pressing the button

"Nothing for you to do" is the sentence that does the damage — it tells you to wait for something that was never going to arrive, and you would have had no way to tell the difference between a slow queue and a dead end. I have filed the underlying problem as **#1017**, because the deeper issue is that this repo has no way to re-evaluate an open PR when `main` moves, and no maintainer-side action that does it without writing to a contributor's branch. Sorry for the noise.

**Review by MakazhanAlpamys — CHANGES_REQUESTED** [review]
Thanks, @Rayan-and-beyond, for the push. It cleared the frozen merge ref, and adding `loss_summary.py` to `TRAINER_MODULES` was what the #755 guard needed. You reused the shared helper instead of writing a third copy, which keeps `initial_loss` / `final_loss` numeric for sweep ranking (#960 criterion 4).

Running `UnlearnTrainerWrapper.train()` with a stubbed per-step loss on the tree merged with current `main`:
- **0 steps:** `main` prints `0.0000`, this PR prints `unavailable`. That is the fix.
- **3 steps:** `2.1000 -> 0.9000` on both, which is correct.
- **1 step:** `main` prints `Loss: 2.1000`, but this PR prints **`Loss: 2.1000 -> 2.1000`** as a delta. That is the fake-delta symptom #899 removed.

Three asks:
1. **The one-step regression.** `src/soup_cli/trainer/unlearn.py:325` passes `[initial_loss, final_loss]`, which are the same step when `step == 1`. Pass only `[final_loss]` when there is a single step, so the helper reports `single`.
2. **No behavioural test protects either fix.** I put the bug back inside the helper calls themselves, `(losses or [0.0])` in `unlearn.py` and `(captured.get("losses") or [0.0])` in `mlx_sft.py`. All 148 related tests pass while the run prints `Loss: 0.0000` again. The exact reverts are caught only by the AST structure guard. Please add tests that call the real code paths:
   - **unlearn:** `UnlearnTrainerWrapper.train()` with a stubbed step loss for 0, 1 and 3 steps, asserting `unavailable`, `single` and `delta`.
   - **MLX:** the result-building path with `captured["losses"]` set to `[]`, `[x]` and `[x, y]`.

   That is #960's third acceptance criterion.
3. **Rebase and credit.** #1018 added `soup_cli/trainer/rewind_mlx.py` on the same line of `TRAINER_MODULES[("sft", "mlx")]`, so `src/soup_cli/config/backend_support.py` now conflicts. Keep both entries. With that resolution I got 1061 passed. Also, `changelog.d/0.75.0/986.fixed.md` should credit `(#960 by @Rayan-and-beyond in #986)`.

**Comment by Rayan-and-beyond** [comment]
Addressed all three review points: one-step unlearn now reports a single loss, both Unlearn and MLX have behavioral 0/1/many-loss coverage, current main is rebased with both rewind_mlx.py and loss_summary.py in the registry, and the changelog credit is updated. The two old zero-fallback mutations now fail the new tests.

**Comment by MakazhanAlpamys** [comment]
Unblocked — I updated your branch, and the red cell was never your code.

`test (windows-latest, 3.10)` was dying with `Windows fatal exception: code 0xc000001d` inside `tests/test_rewind_hf.py` (issue #382): part of GitHub's `windows-latest` fleet lacks an instruction the wheel emits, so the interpreter dies with a faulthandler dump and no Python exception, and everything the cell had not yet reached goes unreported. That has nothing to do with the MLX and unlearn loss reporting you changed.

#1059 merged an hour ago and puts those eight real-training tests behind the shared `skip_on_windows_ci` guard, which fires only on `win32` **and** `CI=true`. I ran `gh pr update-branch` so your PR picks it up; CI is re-running.

Nothing owed on your side — I will merge on green.

**Comment by MakazhanAlpamys** [comment]
Merged — thank you, and thank you for the patience: this sat behind a red cell that was never yours, and the delay was mine.

The defect is small in code and not small in kind. `Loss: 0.0000` where nothing was measured means a run that measured nothing looks exactly like a run whose loss genuinely reached zero — and zero is a number people act on. A placeholder that is indistinguishable from a real result is worse than no number at all, which is why this was worth doing properly rather than papering over with a different default.

You also addressed all three review points and rebased without being chased twice, which I appreciate more than the diff.

Verified before merging, on the tree rebuilt by merging your branch into current `main` (GitHub's merge ref is pinned at run creation, so the green marks described a `main` many merges behind), base interpreter, peft 0.20.0, `soup_cli.__file__` under the scratch tree:

- `ruff check --select F821 src/soup_cli/` — clean.
- The MLX, unlearn and loss-reporting suites on the merged tree — pass.
- A decorator-level diff of the merged tree against `main` — nothing lost. I added that check today after finding that a conflict resolution can silently drop a `@model_validator` line with `ruff` and the import both staying happy.

On the wait: your `test (windows-latest, 3.10)` was dying on #382's illegal-instruction crash inside `tests/test_rewind_hf.py`, a file you do not touch. #1059 put those eight real-training tests behind the shared `win32 AND CI=true` guard, and I ran `update-branch` so this picked it up.

`#960` closes with this.
